Allow setting ``auto_format="True"`` on tool ``output``s.

jmchilton opened the pull request:
The metadata system will then "sniff" the outputs and set the type appropriately. This will be good for tools like upload or genomespace importer where the output type is not known ahead of time - similar tools should no longer need to use wrappers that import galaxy internals in the future.

Tool used to demonstrate and test this functionality is included, run the test with:

```
./run_tests.sh -framework -id output_auto_format
```

See also:
- https://trello.com/c/FZ67pAAE
- http://dev.list.galaxyproject.org/Multiple-outputs-of-unknown-types-td4666891.html
- https://github.com/jmchilton/galaxy/pull/1

Ping @blankenberg

jmchilton committed:
Allow setting ``auto_format="True"`` on tool ``output``s.

The metadata system will then "sniff" the outputs and set the type appropriately.

Tool used to demonstrate and test this functionality is included, run the test with:

    ./run_tests.sh -framework -id output_auto_format

If auto_format="true" than explicitly setting ``format="another_type"`` will prevent the tool from loading so that we can allow this to specify a default type later without breaking backward compatiblity.

Update 1:
 - Switched from ``format="auto"`` to ``auto_format="true"`` based on Dan's comments (https://github.com/galaxyproject/galaxy/pull/130).

Update 2:
 - Updated this to work when setting metadata internally.
 - Switch underlying representation of type from ``_auto_`` to ``_sniff_`` based on seeing a new block in ``galaxy.jobs``. ``auto`` indicates the tool should specify the datatype at runtime. This is in some ways the inverse of that (the tool telling Galaxy to detect the type instead of Galaxy telling the tool to) so I decided it would be best to not have ``auto`` and ``_auto_`` doing completely different things.

guerler commented:
+1. Although I would suggest to name the tag just 'auto'.

jmchilton commented:
Hmm... I would defend `_auto_` in two ways - I think it makes it clearer to developers reading the tools that this field value is special and different than normal formats and it also means there can still exist an `auto` datatype. I am not sure there is an `auto` datatype out there anywhere but I am also not sure there isn't or won't be.

I would have no objection to changing it to `__auto__` or `*auto*` or adding a new field like `auto_format="true"` that does this magic under the hood.

guerler commented:
I suggested this for consistency since the tag is named 'auto' in the upload tool. My second preference would be `__auto__`.

jmchilton commented:
Ahh - so an `auto` datatype wouldn't really work out so well then anyway. Alright I can live with that - thanks for the suggestion.

blankenberg commented:
Perhaps something like `auto_format="true"` with allowing `format="default_type"` to set the default/unsniffable type?

Outputs in worflows will need to be made flexible to be able to connect to inputs that might not match under these conditions.

It is worth noting that the upload tool and the data source tools currently use `format="auto"` for these purposes, but it is mostly handled internally as part of the tool script.

jmchilton commented:
@blankenberg I am not opposed to those semantics - but before adding the complexity - are there any examples of an auto detect tool that would default to anything except `data`?

jmchilton commented:
@blankenberg I have updated the pull request to use this idiom `auto_format="True"`. I have not implemented the second half about defining a default fallback but I have explicitly disallowed setting both fields so later we can add that functionality without breaking backward compatibility due to undefined behaviors.

dannon commented:
+1 to the proposed change, but the test fails for me:

```
FAIL: test_tool_000000 (functional.test_toolbox.TestForTool_output_auto_format)
output_auto_format ( output_auto_format ) > Test-1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/yoplait/galaxy/test/functional/test_toolbox.py", line 268, in test_tool
    self.do_it( td )
  File "/home/yoplait/galaxy/test/functional/test_toolbox.py", line 67, in do_it
    raise e
JobOutputsError: Dataset metadata verification for [file_ext] failed, expected [txt] but found [_auto_].
```

jmchilton commented:
@dannon Hmm... any exceptions further up in the logs about failing to set metadata?

dannon commented:
Didn't see anything obvious, but I can check again.  Passes for you?

jmchilton commented:
It does pass for me - I'll try a clean checkout later today.

dannon commented:
Ok, I can confirm that it works on your branch (but that branch is ~50 commits behind -dev), and upon merge the test fails.

jmchilton commented:
Indeed - just noticed the same thing. I wonder if @jmchilton's metadata "fixes" broke something. I'll look into this more and respond - thanks for the testing!

jmchilton commented:
@dannon since you have graciously done so much testing for me I have a gift for you - sqlalchemy 1.0 seems to have broken setting metadata externally - Galaxy now always falls back to setting it internally (and without warning - the lack of warning is not new however).

So I guess two action items:
- Get setting metadata externally working again.
- Fix `auto_format` when metadata is set internally.

dannon commented:
Hah!  I should have seen that coming, I knew the upgrade was too quiet.  I'll take a look.

jmchilton commented:
Rebased with two new fixes:
- Updated this to work when setting metadata internally.
- Switch underlying representation of type from `_auto_` to `_sniff_` based on seeing a new block in `galaxy.jobs`. `auto` indicates the tool should specify the datatype at runtime. This is in some ways the inverse of that (the tool telling Galaxy to detect the type instead of Galaxy telling the tool to) so I decided it would be best to not have `auto` and `_auto_` doing completely different/opposite things. Regardless this is an implementation detail that shouldn't be exposed to tool authors.

dannon commented:
+1, looks good.  I'll follow up with the sqlalchemy issue elsewhere.

jmchilton commented:
Awesome - thanks for comments @guerler @dannon @blankenberg and for the merge @dannon.
